### DISCUSSION TITLE
Background uploads

### DISCUSSION THREAD

**Discussion by martyfuhry** [781flyingdutchman/background_downloader#256] [General]
Howdy! Thank you for making this wonderful package and creating such helpful documentation. I had a few questions about this package relating to background execution.

Does this package function well in conjunction with background tasks scheduled by the OS while the phone is inactive? In particular, if I schedule an iOS background fetch operation to run some of my dart code, can I create and enqueue upload tasks?

I was roughly hoping to have a background job run my dart code, and maybe create some [NSURLSessionUploadTask](https://developer.apple.com/documentation/foundation/nsurlsessionuploadtask?language=objc) requests. [This rather old discussion](https://forums.developer.apple.com/forums/thread/10239) suggests doing so, but I don't think this package uses that exact class. Maybe it does something equivalent?

The background code execution happens in a separate isolate, and the app may or may not be actively running in the foreground. Is this a viable package for my use case do you think?

**Comment by 781flyingdutchman** [comment]
Hi, thanks! The downloader indeed uses URLSessions and the associated isupload and download tasks, and therefore should not interfere with other parts of the app doing the same, as you'll be using your own, unique URLSessions identifier with your own config.
That said, a few comments on what you're describing:
* Keep in mind that you need to call the `enqueue` or `upload` methods from the main dart isolate - you can't use the downloader from a Dart background isolate
* I don't believe you'll be running your Dart code as a true background task: the code for your background task would have to be written in Swift to truly run in the background. If you "fire up" the Dart engine from such a background task, then as far as iOS is concerned your app is running in the foreground - even if you use a background isolate (because iOS doesn't see the difference between the main and background isolates). This may not matter, but just be aware that true background execution is not possible with Dart code as far as I understand it.

Hope this helps - let me know if you have other questions.

**Reply by mateen-demah** [reply]
@781flyingdutchman is there a reason why the enqueue/upload methods have been confined to the main isolate? It seems like having the ability to enqueue in a background isolate would be helpful, especially in situations when a lot of download tasks have to be enqueued.

**Comment by 781flyingdutchman** [comment]
There is. Two-way communication between Dart and the native platform can only be done on the main isolate. This is a limitation of flutter and I understand they are working on this, but currently that is the case. To avoid starving the main isolate, I suggest adding a TaskQueue just to pace the enqueues. See the docs.

**Reply by mateen-demah** [reply]
Thanks for the quick response. I did try the TaskQueue but even with a 150 millisecond interval, I still couldn't get rid of the stutters.

**Comment by 781flyingdutchman** [comment]
I understand, and I wish there was a better solution. If you're getting stutters at 150ms interval I wonder if you are perhaps doing too much alongside the enqueue, because 150ms is almost 10 frames and an enqueue doesn't take _that_ much time. Perhaps use the profiler to see where the time is spent.

**Comment by 781flyingdutchman** [comment]
See #321 I am working on pre-start and post-finished callbacks that do work when the app is suspended, so if this is still an issue please check out that discussion.
